### PR TITLE
Adding cloud_Drive support for portworx

### PR DESCRIPTION
**Pull request by AmanChh**
Adding cloud _drive parameters for portworx

Refer :https://cloud.ibm.com/catalog/services/portworx-test

source script_sc.sh (Script for getting storageclass )
```
"ibmc-vpc-block-10iops-tier"
"ibmc-vpc-block-5iops-tier"
"ibmc-vpc-block-general-purpose"
"ibmc-vpc-block-custom"
"ibmc-vpc-block-retain-10iops-tier"
"ibmc-vpc-block-retain-5iops-tier"
"ibmc-vpc-block-retain-general-purpose"
"ibmc-vpc-block-retain-custom"

```

### CODE REVIEW

File: examples/portworx/README.md [hunk 1]
```diff
@@ -59,10 +59,14 @@ module "portworx" {
 
 Run the following commands to execute the TF script (containing the modules to create/use ROKS and Portworx). Execution may take about 5-15 minutes:
 
+1)Export IC_API_KEY="XXXXX"
```
Comment (sagarkrdev) [inline examples/portworx/README.md:62]: Can we please add the cloud drive related variables under **Input Variables** section, also please do explain some points on the cloud drive support in readme that what variables become optional when cloud drive is enabled like `install_storage` variable will be always set to false and can't be set to true.

File: examples/portworx/main.tf [hunk 2]
```diff
@@ -46,7 +46,10 @@ module "portworx" {
   unique_id        = var.unique_id
   kube_config_path = data.ibm_container_cluster_config.cluster_config.config_file_path
 
-
+  // Adding cloud drives parameters
+  max_storage_node_per_zone=var.max_storage_node_per_zone
```
Comment (sagarkrdev) [inline examples/portworx/main.tf:50]: let us explain the meaning of these variable in Readme

File: examples/portworx/script_sc.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+response=$(curl -s -d '{"parameters": { "cloud_drive": "Yes"}, "region": "us-south"}' -H "Content-Type: application/json" -X POST https://px-osb-test-2.tapas-iks-1.us-east.containers.appdomain.cloud/sclist)
```
Comment (sagarkrdev) [inline examples/portworx/script_sc.sh:2]: add a TODO to update the URL for fetching storage classes here once cloud drive support is pushed to prod, also is the `region` parameter mandatory and if yes then how we determine the value?

Comment (sagarkrdev) [reply examples/portworx/script_sc.sh:2]: can we please add a sample response of running this script in the PR description if not already done

File: examples/portworx/variables.tf [hunk 4]
```diff
@@ -145,4 +145,39 @@ variable secret_type {
...
+##############################################################################
+# Cloud Drive Variables
+##############################################################################
+variable cloud_drive {
+  description = "cloud drive support enabled"
+  type  = string
+  default = "Yes"
+}
+variable num_cloud_drives{
+  description = "No of drives to provisoned(max=3)"
```
Comment (sagarkrdev) [inline examples/portworx/variables.tf:159]: typo, provisoned -> provision

File: examples/portworx/variables.tf [hunk 5]
```diff
@@ -145,4 +145,39 @@ variable secret_type {
...
+  type=number
+  validation {
+    condition=var.num_cloud_drives>0 && var.num_cloud_drives<4
+    error_message = "Number of drives should be maximum 3"
```
Comment (sagarkrdev) [inline examples/portworx/variables.tf:163]: Would it be better to reword this message to `Maximum number of cloud drives that can be provisioned is 3`?

File: examples/portworx/variables.tf [hunk 6]
```diff
@@ -145,4 +145,39 @@ variable secret_type {
...
+  }
+}
+variable storageClassName {
```
Comment (sagarkrdev) [inline examples/portworx/variables.tf:166]: any reason to use camelCase for this variable name as others are not in that format?

File: examples/portworx/variables.tf [hunk 7]
```diff
@@ -145,4 +145,39 @@ variable secret_type {
...
+  description = "Name of the storage class"
```
Comment (sagarkrdev) [inline examples/portworx/variables.tf:167]: Name of the storage class that will be used to provision cloud drives

File: examples/portworx/variables.tf [hunk 8]
```diff
@@ -145,4 +145,39 @@ variable secret_type {
...
+  type = string
+  default = "ibmc-vpc-block-10iops-tier"
+}
+variable cloud_drives_sizes {
+  description = "Size of the cloud drives"
+  type = list(number)
+  validation {
+    condition=length(var.cloud_drives_sizes)>0 && length(var.cloud_drives_sizes)<4
+    error_message = "Number of drives should be maximum 3"
```
Comment (sagarkrdev) [inline examples/portworx/variables.tf:176]: same error message as suggested above

Comment (sagarkrdev) [inline examples/portworx/variables.tf:176]: same error message as suggested above

File: portworx.tf [hunk 9]
```diff
@@ -47,12 +47,11 @@ resource "kubernetes_secret" "etcd" {
 ##################################
 resource "ibm_resource_instance" "portworx" {
   depends_on = [
-    ibm_container_storage_attachment.volume_attach,
```
Comment (sagarkrdev) [inline portworx.tf:50]: do we need to separate the modules to support cloud drive vs non cloud drive as this variable would be required for installation without cloud drive?

**Review by sagarkrdev — CHANGES_REQUESTED** [review]

File: examples/portworx/variables.tf (hunk 6 above)
Comment (AmanChh) [reply examples/portworx/variables.tf:166]: @sagarkrdev We kept the name in accordance with the api.

**Comment by AmanChh** [comment]
source script_sc.sh (Script for getting storageclass )
```
"ibmc-vpc-block-10iops-tier"
"ibmc-vpc-block-5iops-tier"
"ibmc-vpc-block-general-purpose"
"ibmc-vpc-block-custom"
"ibmc-vpc-block-retain-10iops-tier"
"ibmc-vpc-block-retain-5iops-tier"
"ibmc-vpc-block-retain-general-purpose"
"ibmc-vpc-block-retain-custom"

```

File: modules/portworx_with_cloud_drive/script_sc.sh [hunk 10]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (hkantare) [inline modules/portworx_with_cloud_drive/script_sc.sh:1]: is the script required?

File: examples/portworx_with_cloud_drive/script_sc.sh [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (AmanChh) [reply examples/portworx_with_cloud_drive/script_sc.sh:1]: @hkantare User can run this script to obtain the list of storage class and also they can modify the default value of storageclass in input.tfvars

File: modules/portworx_with_cloud_drive/script_sc.sh (hunk 10 above)
Comment (AmanChh) [reply modules/portworx_with_cloud_drive/script_sc.sh:1]: @hkantare User can use this script to get the name of the storageclass

File: examples/portworx_with_cloud_drive/variables.tf [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+  description = "Do you want to create an external_etcd? `True` or `False`"
+}
+
+variable "region" {
+  description = "The region Portworx will be installed in: us-south, us-east, eu-gb, eu-de, jp-tok, au-syd, etc."
+  default     = "us-south"
+}
+
+variable "kube_config_path" {
+  description = "Path to the k8s config file: ex `~/.kube`"
```
[the remark this answers is not in the document]
Comment (AmanChh) [reply examples/portworx_with_cloud_drive/variables.tf:49]: @hkantare Yeah sure

**Review by sagarkrdev — APPROVED** [review]
LGTM!

**Review by hasan4791 — APPROVED** [review]

**Review by hkantare — APPROVED** [review]

**Pull request merged**